### PR TITLE
Feature(as-types): improved SafeMath's assertions safety

### LINKED ISSUE #215: Improve SafeMath safety with correct assertions

### PR DESCRIPTION
**Pull request by 0xSwapFeeder**
Aim's to resolve #215 from meta #205.

### CODE REVIEW

**Review by Ben-Rey — APPROVED** [review]

File: packages/as-types/assembly/safeMath.ts [hunk 1]
```diff
@@ -26,8 +26,11 @@ export class SafeMath {
    * @throws if the operation results in a number bigger than u64.MAX_VALUE.
    */
   static add(a: u64, b: u64): u64 {
+    assert(a + b >= a, 'SafeMath: addition overflow');
+    assert(a + b >= 0, 'SafeMath: addition overflow');
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:30]: useless, a and b are unsigned so by design this assertion will always be true

File: packages/as-types/assembly/safeMath.ts [hunk 2]
```diff
@@ -26,8 +26,11 @@ export class SafeMath {
...
+    assert(a + b <= u64.MAX_VALUE, 'SafeMath: addition overflow');
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:31]: same as above, by design this is always true

File: packages/as-types/assembly/safeMath.ts [hunk 3]
```diff
@@ -26,8 +26,11 @@ export class SafeMath {
    * @throws if the operation results in a number bigger than u64.MAX_VALUE.
    */
   static add(a: u64, b: u64): u64 {
+    assert(a + b >= a, 'SafeMath: addition overflow');
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:29]: This is the real assertion, but not well presented IMHO.
What you want is that:
a + b < u64.MAX_VALUE
<=> b < u64.MAX_VALUE - a (we substract by a on the left and on the right of the inequality as a >= 0)

So you assertiion should be :
b < u64.MAX_VALUE - a
```suggestion
    assert(b < u64.MAX_VALUE - a, 'SafeMath: addition overflow');
```

File: packages/as-types/assembly/safeMath.ts [hunk 4]
```diff
@@ -68,12 +71,12 @@ export class SafeMath {
    * @throws if the operation results in a number bigger than u64.MAX_VALUE.
    */
   static mul(a: u64, b: u64): u64 {
-    if (a == 0) {
+    if (a == 0 || b == 0) {
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:74]: This update is useless: the check of the null value is not for performance, but to divide by a.

File: packages/as-types/assembly/safeMath.ts [hunk 5]
```diff
@@ -68,12 +71,12 @@ export class SafeMath {
...
       return 0;
     }
+    assert((a * b) / a == b, 'SafeMath: multiplication overflow');
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:77]: Can you correct this assertion using what I did for the previous one ?

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: packages/as-types/assembly/safeMath.ts [hunk 6]
```diff
@@ -26,9 +26,7 @@ export class SafeMath {
    * @throws if the operation results in a number bigger than u64.MAX_VALUE.
    */
   static add(a: u64, b: u64): u64 {
-    assert(a + b >= a, 'SafeMath: addition overflow');
-    assert(a + b >= 0, 'SafeMath: addition overflow');
-    assert(a + b <= u64.MAX_VALUE, 'SafeMath: addition overflow');
+    assert(b < u64.MAX_VALUE - a, 'SafeMath: addition overflow');
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:29]: b = u64.MAX_VALUE and a = 0 shall pass
```suggestion
    assert(b <= u64.MAX_VALUE - a, 'SafeMath: addition overflow');
```

File: packages/as-types/assembly/safeMath.ts [hunk 7]
```diff
@@ -71,9 +72,9 @@ export class SafeMath {
     if (a == 0) {
       return 0;
     }
+    assert(b < u64.MAX_VALUE / a, 'SafeMath: multiplication overflow');
```
Comment (gregLibert) [inline packages/as-types/assembly/safeMath.ts:75]: You start with:
a*b <= u.MAX
<=> b <= u.max/a, a > 0

```suggestion
    assert(b <= u64.MAX_VALUE / a, 'SafeMath: multiplication overflow');
```

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Review by gregLibert — APPROVED** [review]

**Pull request merged**